prow/cmd/mkbuild-cluster/main.go

```diff
@@ -0,0 +1,331 @@
+	zone          string
+}
+
+type describe struct {
+	Auth     describeAuth `json:"masterAuth"`
+	Endpoint string       `json:"endpoint"`
+}
+
+type describeAuth struct {
+	ClientCertificate    string `json:"clientCertificate"`
+	ClientKey            string `json:"clientKey"`
+	ClusterCACertificate string `json:"clusterCaCertificate"`
+}
+
+func parseOptions() options {
+	var o options
+	if err := o.parseArgs(flag.CommandLine, os.Args[1:]); err != nil {
+		logrus.Fatalf("Invalid flags: %v", err)
+	}
+	return o
+}
+
+func (o *options) parseArgs(flags *flag.FlagSet, args []string) error {
+	flags.StringVar(&o.account, "account", "", "use this account to describe --cluster")
+	flags.StringVar(&o.alias, "alias", "", "the --build-cluster alias to add")
+	flags.StringVar(&o.cluster, "cluster", "", "the GKE cluster to describe")
+	flags.StringVar(&o.project, "project", "", "the GKE project to describe")
+	flags.StringVar(&o.zone, "zone", "", "the GKE zone to describe")
+	flags.BoolVar(&o.printData, "print-file", false, "print the file outside of the configmap secret")
+	flags.BoolVar(&o.printEntry, "print-entry", false, "print the new entry without appending to existing ones at stdin")
+	flags.BoolVar(&o.getClientCert, "get-client-cert", false, fmt.Sprintf("first get-credentials for the cluster using %s=True", useClientCertEnv))
+	flags.BoolVar(&o.changeContext, "change-context", false, "allow --get-client-cert to change kubectl config current-context")
+	flags.BoolVar(&o.skipCheck, "skip-check", false, "skip validating the creds work in a client")
+	switch err := flags.Parse(args); {
+	case err != nil:
+		return err
+	case o.cluster == "":
+		return errors.New("--cluster required")
+	case o.project == "":
+		return errors.New("--project required")
+	case o.zone == "":
+		return errors.New("--zone required")
+	case o.alias == "":
+		return fmt.Errorf("--alias required (use %q for default)", kube.DefaultClusterAlias)
+	}
+	return nil
+}
+
+func main() {
+	// Gather options from flags
+	o := parseOptions()
+	if err := do(o); err != nil {
+		logrus.Fatalf("Failed: %v", err)
+	}
+}
+
+// useContext calls kubectl config use-context ctx
+func useContext(o options, ctx string) error {
+	_, cmd := command("kubectl", "config", "use-context", ctx)
+	return cmd.Run()
+}
+
+// currentContext returns kubectl config current-context
+func currentContext(o options) (string, error) {
+	_, cmd := command("kubectl", "config", "current-context")
+	b, err := cmd.Output()
+	return strings.TrimSpace(string(b)), err
+}
+
+// getCredentials calls gcloud container clusters get-credentials, usually preserving currentContext()
+func getCredentials(o options) error {
+	if !o.changeContext {
+		cur, err := currentContext(o)
+		if err != nil {
+			return fmt.Errorf("read current-context: %v", err)
+		}
+		defer useContext(o, cur)
+	}
+
+	// TODO(fejta): we ought to update kube.Client to support modern auth methods.
+	// More info: https://github.com/kubernetes/kubernetes/issues/30617
```

Comment:
Yeah, we really should be using GCP service account credentials which have better management (revocation etc.), rather than the legacy auth stuff... roughly outlined in https://github.com/kubernetes/kubernetes/issues/30617#issuecomment-240852122

Probably the sanest and most portable route long term imho is:
- use the upstream client ([client-go](https://github.com/kubernetes/client-go))
- just [read in a `KUBECONFIG`](https://github.com/kubernetes/client-go/blob/master/examples/out-of-cluster-client-configuration) file contents for each build cluster...

This is portable, and maintained upstream. If we abstracted over a `KUBECONFIG` per build cluster you'd be able to just say "get a working kubeconfig for your cluster, and put it in a secret". For building against the same cluster [we can also support an in-cluster](https://github.com/kubernetes/client-go/tree/master/examples/in-cluster-client-configuration) kubernetes service account token.

Reply:
Yes, refactoring/replacing prow/kube is for another PR